# scribe

This repository contains our implementation for the paper *Inferring Causal Gene Regulatory Networks from Coupled Single-Cell Expression Dynamics Using Scribe* accepted by Cell Systems.

The paper online version can be found [here](https://www.sciencedirect.com/science/article/pii/S2405471220300363). 

The official implementation and its variants can be found [here](https://github.com/aristoteleo/Scribe-py/tree/master).

## Requirements

To configure the environment, you can create a conda environment and install the `environment.yml` by:

```shell
conda env create -f environment.yml --name scribe
```

Our environment included:

- numpy=1.23.5
- pandas=1.5.2
- python=3.9.16
- pip:
  - scikit-learn==1.2.1
  - scipy==1.10.0
  - tqdm==4.64.1

## Run experiments

In general, following args are used to select the trajectories to be used for evaluation:

```R
parser.add_argument('--data-path', type=str,
                    default="/work/projects/bsimds/backup/src/simulations/",
                    help="The folder where data are stored.")
parser.add_argument('--save-folder', type=str, required=True,
                    help="The folder where resulting adjacency matrixes are stored.")
parser.add_argument('--b-portion', type=float, default=1.0,
                    help='Portion of data to be used in benchmarking.')
parser.add_argument('--b-time-steps', type=int, default=49,
                    help='Portion of time series in data to be used in benchmarking.')
parser.add_argument('--b-shuffle', action='store_true', default=False,
                    help='Shuffle the data for benchmarking?')
parser.add_argument('--b-network-type', type=str, default='',
                    help='What is the network type of the graph.')
parser.add_argument('--b-directed', action='store_true', default=False,
                    help='Default choose trajectories from undirected graphs.')
parser.add_argument('--b-simulation-type', type=str, default='',
                    help='Either springs or netsims.')
parser.add_argument('--b-suffix', type=str, default='',
		            help='The rest to locate the exact trajectories. E.g. "50r1_n1" for 50 nodes, rep 1 and noise level 1. Or "50r1" for 50 nodes, rep 1 and noise free.')
parser.add_argument('--pct-cpu', type=float, default=1.0,
                    help='Percentage of number of CPUs to be used.')
```

Reproduce the results of scribe in the noise-free trajectories generated by NetSims simulation, and by Brain Networks with 15 nodes, with the first repetition number:

```shell
python run.py --save-folder="./results" --b-network-type="brain_networks" --b-directed --b-simulation-type="netsims" --b-suffix="test_netsims15r1.npy" &
```

Reproduce the results of scribe in the noise-free trajectories generated by NetSims simulation, and by Brain Networks with 30 nodes, with the second repetition number:

```shell
python run.py --save-folder="./results" --b-network-type="brain_networks" --b-directed --b-simulation-type="netsims" --b-suffix="test_netsims30r2.npy" &
```

Reproduce the results of scribe in the noisy trajectories generated by NetSims simulation, and by Brain Networks with 50 nodes, with the third repetition number, with two levels of added Gaussian noise:

```shell
python run.py --save-folder="./results" --b-network-type="brain_networks" --b-directed --b-simulation-type="netsims" --b-suffix="test_netsims50r3_n2.npy" &
```

Reproduce the results of scribe in the noise-free trajectories generated by NetSims simulation, by Brain Networks with 30 nodes, with the second repetition number, and with 5 time steps:

```shell
python run.py --save-folder="./results" --b-network-type="brain_networks" --b-directed --b-simulation-type="netsims" --b-suffix="test_netsims30r2.npy" --b-time-steps=5 &
```

